src/string.rs

```diff
@@ -0,0 +1,171 @@
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Format comments and string literals
+
+use utils::*;
+
+use MIN_STRING;
+
+pub struct StringFormat<'a> {
+    pub opener: &'a str,
+    pub closer: &'a str,
+    pub line_start: &'a str,
+    pub line_end: &'a str,
+    pub width: usize,
+    pub offset: usize,
+    pub trim_end: bool
+}
+
+// TODO: take starting column number -- we could do a lot with that
+pub fn rewrite_comment(orig: &str, block_style: bool, width: usize, offset: usize) -> String {
+    let s = orig.trim();
+
+    // Edge case: block comments. Let's not trim their lines (for now)
+    let opener = if block_style { "/* " } else { "// " };
+    let closer = if block_style { " */" } else { "" };
+    let line_start = if block_style { "* " } else { "// " };
+
+    let max_chars = width.checked_sub(closer.len()).unwrap_or(1)
+                         .checked_sub(opener.len()).unwrap_or(1);
+
+    let fmt = StringFormat {
+        opener: "",
+        closer: "",
+        line_start: line_start,
+        line_end: "",
+        width: max_chars,
+        offset: offset + opener.len() - line_start.len(),
+        trim_end: true
+    };
+
+    let indent_str = make_indent(offset + opener.len() - line_start.len());
+
+    let (_, mut s) = s.lines()
+                  .map(|line| {
+                      // Remove first whitespace, because we'll add that extra later
+                      line.trim()
+                          .trim_left_matches("// ")
+                          .trim_left_matches("//")
+                          .trim_left_matches("/* ")
+                          .trim_left_matches("/*")
+                          .trim_left_matches("* ")
+                          .trim_left_matches('*')
+                          .trim_right_matches("*/")
+                          .trim_right_matches(' ')
+                  })
+                  .fold((true, opener.to_owned()), |(first, mut a), x| {
+                      if !first {
+                          a.push('\n');
+                          a.push_str(&indent_str);
+                          a.push_str(line_start);
+                      }
+
+                      if x.len() > max_chars {
+                          a.push_str(&rewrite_string(x, &fmt));
+                      } else {
+                          a.push_str(x);
+                      }
+
+                      (false, a)
+                  });
+
+    s.push_str(closer);
+
+    // Trim inner comment when it fits on one line
+    if s.len() < width && !s.contains('\n') {
+        s.trim().to_owned()
+    } else {
+        s
+    }
```

Comment:
In which case is this needed ? It seems to me that `s` is already trimmed since all lines are individually trimmed.